cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataSubscriberService.java

```diff
@@ -96,114 +95,104 @@
         NamespaceId.SYSTEM, ImmutableMap.of(), null, null, messagingContext)),
       org.apache.tephra.RetryStrategies.retryOnConflict(20, 100)
     );
-
-    // Adding all metadata related subscribers
-    this.subscriberServices = Collections.singletonList(new DataAccessLineageSubscriber());
   }
 
   /**
    * Sets the {@link DatasetId} for the {@link LineageDataset}. This method is only for testing.
    */
   @VisibleForTesting
-  public void setLineageDatasetId(DatasetId lineageDatasetId) {
+  void setLineageDatasetId(DatasetId lineageDatasetId) {
     this.lineageDatasetId = lineageDatasetId;
   }
 
   @Override
-  protected void startUp() throws Exception {
-    // Start all subscriber services. All of them has no-op in start, so they shouldn't fail.
-    Futures.successfulAsList(subscriberServices.stream().map(s -> {
-      LOG.debug("Starting metadata subscriber {}", s.getClass().getSimpleName());
-      return s.start();
-    }).collect(Collectors.toList())).get();
-
-    LOG.debug("All metadata subscribers started");
+  protected MessagingContext getMessagingContext() {
+    return messagingContext;
+  }
+
+  @Override
+  protected Transactional getTransactional() {
+    return transactional;
+  }
+
+  @Override
+  protected MetadataMessage decodeMessage(Message message) throws Exception {
+    return GSON.fromJson(message.getPayloadAsString(), MetadataMessage.class);
+  }
+
+  @Nullable
+  @Override
+  protected String loadMessageId(DatasetContext datasetContext) throws Exception {
+    AppMetadataStore appMetadataStore = AppMetadataStore.create(cConf, datasetContext, datasetFramework);
+    return appMetadataStore.retrieveSubscriberState(getTopicId().getTopic());
   }
 
   @Override
-  protected void shutDown() throws Exception {
-    // This never throw
-    Futures.successfulAsList(subscriberServices.stream().map(s -> {
-      LOG.debug("Stopping metadata subscriber {}", s.getClass().getSimpleName());
-      return s.stop();
-    }).collect(Collectors.toList())).get();
-
-    for (Service service : subscriberServices) {
-      // The service must have been stopped, and calling stop again will just return immediate with the
-      // future that carries the stop state.
-      try {
-        service.stop().get();
-      } catch (ExecutionException e) {
-        LOG.warn("Exception raised when stopping service {}", service.getClass().getSimpleName(), e.getCause());
+  protected void storeMessageId(DatasetContext datasetContext, String messageId) throws Exception {
+    AppMetadataStore appMetadataStore = AppMetadataStore.create(cConf, datasetContext, datasetFramework);
+    appMetadataStore.persistSubscriberState(getTopicId().getTopic(), messageId);
```

Comment:
this method assumes there's one subscriber per topic. Do you think it makes sense to use a subscriber name instead of the topic name (and change the argument name in AppMetadataStore)?

Reply:
I think it could be some prefix (like subscriber-id) plus the topic name. Ideally we'll like to be able to handle cases when topic changed for the same subscriber-id.

Reply:
@albertshau I updated the API to take both topic name and subscriber id in the last commit. Please take a look.